solidity/ecdsa/contracts/libraries/EcdsaAuthorization.sol

```diff
@@ -15,25 +15,474 @@
+    ///         needs to update state of the sortition pool with a call to
+    ///         `joinSortitionPool` or `updateOperatorStatus`. After the
+    ///         sortition pool state is in sync, authorization decrease delay
+    ///         starts.
+    ///
+    ///         After authorization decrease delay passes, authorization
+    ///         decrease request needs to be approved with a call to
+    ///         `approveAuthorizationDecrease` function.
+    ///
+    ///         If there is a pending authorization decrease request, it is
+    ///         overwritten.
+    ///
+    /// @dev Should only be callable by T staking contract.
+    function authorizationDecreaseRequested(
+        Data storage self,
+        address stakingProvider,
+        uint96 fromAmount,
+        uint96 toAmount
+    ) internal {
+        require(
+            toAmount == 0 || toAmount >= self.parameters.minimumAuthorization,
+            "Authorization amount should be 0 or above the minimum"
+        );
+
+        address operator = self.stakingProviderToOperator[stakingProvider];
+
+        uint64 decreasingAt;
+
+        if (operator == address(0)) {
+            // Operator is not known. It means `registerOperator` was not
+            // called yet, and there is no chance the operator could
+            // call `joinSortitionPool`. We can let to approve authorization
+            // decrease immediately because that operator was never in the
+            // sortition pool.
+
+            // solhint-disable-next-line not-rely-on-time
+            decreasingAt = uint64(block.timestamp);
+        } else {
+            // Operator is known. It means that this operator is or was in
+            // the sortition pool. Before authorization decrease delay starts,
+            // the operator needs to update the state of the sortition pool
+            // with a call to `joinSortitionPool` or `updateOperatorStatus`.
+            // For now, we set `decreasingAt` as "never decreasing" and let
+            // it be updated by `joinSortitionPool` or `updateOperatorStatus`
+            // once we know the sortition pool is in sync.
+
+            // solhint-disable-next-line not-rely-on-time
+            decreasingAt = type(uint64).max;
+        }
+
+        uint96 decreasingBy = fromAmount - toAmount;
+
+        self.pendingDecreases[stakingProvider] = AuthorizationDecrease(
+            decreasingBy,
+            decreasingAt
+        );
+
+        emit AuthorizationDecreaseRequested(
+            stakingProvider,
+            operator,
+            fromAmount,
+            toAmount,
+            decreasingAt
+        );
+    }
+
+    /// @notice Approves the previously registered authorization decrease
+    ///         request. Reverts if authorization decrease delay have not passed
+    ///         yet or if the auhorization decrease was not requested for the
+    ///         given staking provider.
+    function approveAuthorizationDecrease(
```

Comment:
There are two possible paths.

The first path is that a new authorization decrease request overwrites a pending one in T `TokenStaking`. Thus, it is possible to request authorization decrease by 0. 

In this case, `TokenStaking` will call `EcdsaAuthorization` with `deauthorizingTo` equal to [the entire authorized amount](https://github.com/threshold-network/solidity-contracts/blob/702aa5f12242e885e7b0c2b3228bb788439d30a8/contracts/staking/TokenStaking.sol#L1314). As a result, `decreasingBy` [will evaluate to 0](https://github.com/keep-network/keep-core/blob/68f072cdedc6159e75c351ec8a1bba2694ccf74a/solidity/ecdsa/contracts/libraries/EcdsaAuthorization.sol#L240) and will overwrite the pending `self.pendingDecreases[stakingProvider]`. `isOperatorUpToDate` [should start reporting `false`](https://github.com/keep-network/keep-core/blob/68f072cdedc6159e75c351ec8a1bba2694ccf74a/solidity/ecdsa/contracts/libraries/EcdsaAuthorization.sol#L450-L454) and the sortition pool state should be updated.

The second path is that someone increases their authorization using the remaining portion of their stake while having a pending authorization decrease.

This is covered by `EcdsaAuthorization` as well because `eligibleStake` should [start returning a higher amount](https://github.com/keep-network/keep-core/blob/68f072cdedc6159e75c351ec8a1bba2694ccf74a/solidity/ecdsa/contracts/libraries/EcdsaAuthorization.sol#L472-L475), and `isOperatorUpToDate` should [start returning false](https://github.com/keep-network/keep-core/blob/68f072cdedc6159e75c351ec8a1bba2694ccf74a/solidity/ecdsa/contracts/libraries/EcdsaAuthorization.sol#L450) and the sortition pool should get updated accordingly.